capsules/src/net/sixlowpan.rs

```diff
@@ -11,91 +11,267 @@
-//! its TxState struct, and clients should not directly modify any fields within
-//! the TxState struct. The fragmentation layer uses this struct to store state
-//! about packets currently being sent.
 //!
-//! RxState:
-//! - Methods:
-//! -- new(..): Initializes a new RxState struct
+//! The interface is explored in more detail below.
 //!
-//! The RxState struct contains information relating to packet reception and
-//! reassembly. Unlike with the TxState structs, there is no concept of
-//! individual clients "owning" these structs; instead, some number of RxStates
-//! are statically allocated and added to the FragState's RxState list. These
-//! states are then used to keep track of different reassembly flows; the
-//! number of simultaneous packet receptions is dependent on the number of
-//! allocated RxState structs. When a packet is fully reassembled, the global
-//! receive client inside FragState receives a callback.
+//! *Sixlowpan Struct:* For the `Sixlowpan` struct, we divide the
+//! user interface into two parts: standard usage and initialization. First,
+//! for standard usage (transmission), the `Sixlowpan` struct supplies the
+//! following method:
 //!
-//! In addition to structs and their methods, this layer also defines several
-//! traits for the transmit and receive callbacks.
+//! ```
+//! Sixlowpan::transmit_packet(&self,
+//!                            src_mac_addr: MacAddress,
+//!                            dst_mac_addr: MacAddress,
+//!                            ip6_packet: &'static mut [u8],
+//!                            ip6_packet_len: usize,
+//!                            security: Option<(SecurityLevel, KeyId)>,
+//!                            fragment: bool,
+//!                            compress: bool)
+//!                            -> Result<ReturnCode, ReturnCode>;
+//! ```
+//! This function exposes the primary packet transmission fuctionality. The
+//! source and destination mac address arguments specify the link-layer
+//! addresses of the packet, while the `security`, `fragment`, and `compress`
+//! options specify the security level (if any), whether to fragment the packet
+//! if it is too large, and whether to compress the packet respectively. These
+//! different options are exposed to provide the caller with more control over
+//! how this layer modifies packets; no compression and no fragmentation can
+//! be set, and the packet will be sent to the radio as a raw IPv6 packet.
 //!
-//! TransmitClient Trait:
-//! - send_done(..): Called after the entire IPv6 packet has been sent. Returns
-//!     the IPv6 buffer, a reference to the TxState struct, and additional
-//!     information relating to the sent packet
+//! The `ip6_packet` argument contains a pointer to a buffer containing a valid
+//! IPv6 packet, while the `ip6_packet_len` argument specifies the number of
+//! bytes to send. Note that `ip6_packet.len() > ip6_packet_len`, but we check
```

Comment:
Right, so this touches on a slightly problematic part of Rust - that once you pass ownership of a mutable slice to a different layer, you cannot get the full buffer back. So in the transmit case, we need to be able to pass the full `ip6_packet` buffer to the function (and receive it back in a callback), but the packet may only take up a subsection of this buffer. So the statement `ip6_packet.len()` simply returns the full length of the buffer, but you must pass in a length that is *less* than that buffer (to prevent us indexing past the end).

So although a little confusing, this just states that the buffer you pass in *can* be longer than the length, but the length *must* be less than or equal to the total buffer length. The comment should probably be changed to make this more clear.

Reply:
So basically you are saying that the line "Note that 'ip6_packet.len() > ip6_packet_len'" Really should read "Note that 'ip6_packet.len() must be >= ip6_packet_len'"? Or am I still misunderstanding?

Reply:
Yeah. I can fix the comment to make this more apparent.